### PR TITLE
[Beat] Refactor E-KYC Solution API for 1x Hit

### PR DESCRIPTION
**Pull request by lazuardyk**
## Type of changes

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Enhancement
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe): 

## What is the current behaviour?
- e-KYC solution is 3x hit from front end
- using ```/services``` endpoint

## What is the new behavior?
- e-KYC solution is 1x hit from front end
- using ```/ekyc``` endpoint

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Other information

Screenshot from Postman:
Screenshot_1891

Note that this endpoint will take more time than usual because we did a 3 requests to the cloud **sequentially**.

### CODE REVIEW

**Comment by sonadztux** [comment]
Jadi berapa lama @lazuardyk?

File: backend/README.md [hunk 1]
```diff
@@ -227,7 +236,182 @@ OR
 </details>
 
 <details>
-<summary><b>Create A Service Request By ID</b></summary>
+<summary><b>Create A Service Request for e-KYC [Solution]</b></summary>
+Create a service request for E-KYC solution.
+
+- **URL**
+
+  `/ekyc`
```
Comment (sonadztux) [inline backend/README.md:244]: Zu minta tolong ini disesuain langsung biar jadi `/api/v1/ekyc` boleh ga?

File: backend/README.md [hunk 2]
```diff
@@ -227,7 +236,182 @@ OR
...
+
+- **Method**
+
+  `POST`
+
+- **Request Payload**
+
+```json
+{
```
Comment (sonadztux) [inline backend/README.md:253]: `additional_params` kayaknya ga perlu jadi payload kalo emang isinya kosong

File: backend/routes.go [hunk 3]
```diff
@@ -38,5 +38,10 @@ func SetupRouter() *gin.Engine {
 		feedbacks.POST("/:service_id", controllers.CreateFeedback)
 	}
 
+	ekyc := r.Group("/ekyc")
+	{
+		ekyc.POST("", controllers.GetEKYCResult)
```
Comment (sonadztux) [inline backend/routes.go:43]: Ini nama controllernya jadiin `CreateEKYCRequest` aja biar konsisten sama yang atasnya, semua nama controller buat request POST pakai prefix `Create`

File: backend/controllers/ekyc.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	Data      ekycRequestData `json:"data"`
+}
+
+type ekycServiceData struct {
+	FaceLiveness models.ServiceRequestResultData `json:"face_liveness"`
+	OCRKTP       models.ServiceRequestResultData `json:"ocr_ktp"`
+	FaceMatch    models.ServiceRequestResultData `json:"face_match"`
+}
+
+func GetEKYCResult(ctx *gin.Context) {
```
Comment (sonadztux) [inline backend/controllers/ekyc.go:30]: Ini namanya ubah jadi `CreateEKYCRequest` gimana?

**Review by sonadztux — CHANGES_REQUESTED** [review]

File: backend/README.md (hunk 2 above)
Comment (rizkybaihaqy) [reply backend/README.md:253]: setuju

File: backend/controllers/ekyc.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+type ekycRequestInput struct {
+	SessionID string          `json:"session_id"`
+	Data      ekycRequestData `json:"data"`
+}
+
+type ekycServiceData struct {
+	FaceLiveness models.ServiceRequestResultData `json:"face_liveness"`
+	OCRKTP       models.ServiceRequestResultData `json:"ocr_ktp"`
+	FaceMatch    models.ServiceRequestResultData `json:"face_match"`
+}
```
Comment (sonadztux) [inline backend/controllers/ekyc.go:28]: Ini pindah ke models/ aja

File: backend/controllers/ekyc.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	fmt.Println(input)
+	var result models.ServiceRequestResultData
+	dataAnalytic := GetDataAnalytic(service, input)
+	result, err := RequestToAnalyticSync(dataAnalytic, "face-liveness")
+	if err != nil {
+		fmt.Println("Error during fetching API face liveness: ", err)
+		return result, err
+	}
+	return result, nil
+}
```
Comment (sonadztux) [inline backend/controllers/ekyc.go:109]: Buat face payment butuh ini juga, enaknya gimana ya biar ga buat baru (duplicate)?

File: backend/controllers/ekyc.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+func getResultFaceMatch(service models.Service, input models.RequestData) (models.ServiceRequestResultData, error) {
+	var result models.ServiceRequestResultData
+	dataAnalytic := GetDataAnalytic(service, input)
+	result, err := RequestToAnalyticSync(dataAnalytic, "face-match")
+	if err != nil {
+		fmt.Println("Error during fetching API face match: ", err)
+		return result, err
+	}
+	return result, nil
+}
```
Comment (sonadztux) [inline backend/controllers/ekyc.go:131]: Face payment butuh ini juga

File: backend/controllers/ekyc.go (hunk 6 above)
Comment (lazuardyk) [reply backend/controllers/ekyc.go:109]: Utk sementara bisa dicopy dulu yak fungsinya, gw mau nyoba yang ubah ke async dulu, gw commit kalo udah selesai

Comment (sonadztux) [reply backend/controllers/ekyc.go:109]: sama itu line 100 hapus aja Println nya

File: backend/controllers/ekyc.go (hunk 5 above)
Comment (krismp) [reply backend/controllers/ekyc.go:28]: `xRequestData` sm `xRequestInput` bedanya apa @lazuardyk ?

File: backend/controllers/ekyc.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+func getResultOCRKTP(service models.Service, input models.RequestData) (models.ServiceRequestResultData, error) {
+	var result models.ServiceRequestResultData
+	dataAnalytic := GetDataAnalytic(service, input)
+	result, err := RequestToAnalyticSync(dataAnalytic, "ocr-ktp")
+	if err != nil {
+		fmt.Println("Error during fetching API ocr ktp: ", err)
```
Comment (krismp) [inline backend/controllers/ekyc.go:116]: kita udh perlu nambahin log kayanya, nanti ini bisa masuk log severity error, gw bikin issue nya ya

Comment (krismp) [reply backend/controllers/ekyc.go:116]: https://github.com/nodefluxio/weber/issues/72

File: backend/README.md (hunk 1 above)
Comment (lazuardyk) [reply backend/README.md:244]: Done https://github.com/nodefluxio/weber/pull/68/commits/a64810572eccc2a1b447c4229fe473539bec5f1d

File: backend/README.md (hunk 2 above)
Comment (lazuardyk) [reply backend/README.md:253]: Done https://github.com/nodefluxio/weber/pull/68/commits/e400e63487cc22534cd929b2e8d1be30a2796676

File: backend/routes.go (hunk 3 above)
Comment (lazuardyk) [reply backend/routes.go]: Done https://github.com/nodefluxio/weber/pull/68/commits/a64810572eccc2a1b447c4229fe473539bec5f1d

File: backend/controllers/ekyc.go (hunk 4 above)
Comment (lazuardyk) [reply backend/controllers/ekyc.go]: Done https://github.com/nodefluxio/weber/pull/68/commits/4359a53f5aa7ac3aadbfdb39d8af98296a3e7866

File: backend/controllers/ekyc.go (hunk 6 above)
Comment (lazuardyk) [reply backend/controllers/ekyc.go]: Ini gw pindahin ke ```analytic.go``` biar bisa dipake bareng
https://github.com/nodefluxio/weber/pull/68/commits/7a6882db18d9ad6612600217c0fd913c27abeaa5

File: backend/controllers/ekyc.go (hunk 7 above)
Comment (lazuardyk) [reply backend/controllers/ekyc.go]: Done https://github.com/nodefluxio/weber/pull/68/commits/7a6882db18d9ad6612600217c0fd913c27abeaa5

File: backend/controllers/ekyc.go (hunk 8 above)
Comment (lazuardyk) [reply backend/controllers/ekyc.go]: Ok boleh, di issue dulu ya

File: backend/controllers/ekyc.go (hunk 5 above)
Comment (lazuardyk) [reply backend/controllers/ekyc.go]: @sonadztux terkait ini gaada hubungannnya sama database langsung sih, ditaro models aja?
sbnrnya yg di analytic.go juga ada 1 struct. tbh gw mikirnya jadi ribet soalnya kalo pindah2 file ke models wkwk

@krismp ```RequestInput``` itu ngebungkus ```RequestData```. karena lemparan dari frontend itu ada session id dan data

**Comment by lazuardyk** [comment]
Udah gw update jadi async & resolve conflict. boleh dicoba dan dire-review ya @sonadztux @krismp @rizkybaihaqy

File: backend/controllers/ekyc.go (hunk 6 above)
Comment (sonadztux) [reply backend/controllers/ekyc.go]: Sip udah gue pake. Thanks @lazuardyk

**Review by sonadztux — APPROVED** [review]

**Review by bagasteguhimani — APPROVED** [review]
Work!

**Pull request merged**